cmd/security-secrets-setup/main_test.go

```diff
@@ -0,0 +1,138 @@
+//
+// Copyright (c) 2019 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+// in compliance with the License. You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software distributed under the License
+// is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+// or implied. See the License for the specific language governing permissions and limitations under
+// the License.
+//
+// SPDX-License-Identifier: Apache-2.0'
+//
+package main
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"testing"
+
+	"github.com/edgexfoundry/edgex-go/internal/security/setup"
+	"github.com/edgexfoundry/edgex-go/internal/security/setup/option"
+	"github.com/stretchr/testify/assert"
+)
+
+type testExitCode struct {
+	testStatusCode int
+}
+
+type testPkiInitOptionDispatcher struct {
+	testOptsExecutor option.OptionsExecutor
+}
+
+var hasDispatchError bool
+
+func TestMainWithNoOption(t *testing.T) {
+	tearDown := setupTest(t)
+	origArgs := os.Args
+	defer tearDown(t, origArgs)
+	assert := assert.New(t)
+
+	runWithNoOption()
+	assert.Equal(1, (exitInstance.(*testExitCode)).getStatusCode())
+}
+
+func TestMainWithConfigFileOption(t *testing.T) {
+	tearDown := setupTest(t)
```

Comment:
this kind of test usually reads much cleaner as:

```go
defer mockExitInstance(newArgs(), newTestExit(), newTestDispatcher())
```

where `mockExitInstance` looks like this:

```go
func mockExitInstance(mockedExiter exit) func() {
    // save for restoring
    prevExitInstance = exitInstance

    // mock the new one
    exitInstance = mockedExiter

    return func() {
        exitInstance = prevExitInstance

        // you can also do cleanup here too
        os.RemoveAll("./config")
    }
}
```

repeating this pattern for all the things you need to mock

Reply:
Well, my original real purpose to do this is just for prepare Initialization/reset and tearDown for unit tests and not really for mocking purpose per se. But your idea to do this way is interesting for mocking things for other testing purpose as well.  The real mocking is actually achieved via combination of interface and testing struct like I do here.